provisionerd/runner.go

```diff
@@ -0,0 +1,872 @@
+package provisionerd
+
+import (
+	"archive/tar"
+	"bytes"
+	"context"
+	"errors"
+	"fmt"
+	"io"
+	"os"
+	"path"
+	"path/filepath"
+	"reflect"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/google/uuid"
+	"github.com/spf13/afero"
+	"golang.org/x/xerrors"
+
+	"cdr.dev/slog"
+
+	"github.com/coder/coder/provisionerd/proto"
+	sdkproto "github.com/coder/coder/provisionersdk/proto"
+)
+
+type jobRunner interface {
+	start()
+	cancel()
+	isDone() <-chan any
+	fail(ctx context.Context, f *proto.FailedJob) error
+	forceStop()
+}
```

Comment:
This interface seems to only be used in a single place. Could we return `*runner` instead?

Reply:
I'm specifically trying to downscope the interface surface, so that you can't call methods that are meant to be internal to the runner, and you can't access data that is protected by mutex, etc.

If you really want to avoid the interface, I could move this to a different package, and then only export these methods, but the resulting package feels a little too small.  Also, I do the same thing with the runner's view of the provisionerd server --- defining an interface so it can't touch internals.  But, we don't want to export those methods from the package.

Reply:
I don't believe the interface will reduce what a future contributor will expose, because the scope of the interface isn't clearly defined. It's a job runner, but would that mean it doesn't acquire its own jobs? I'd think the code should _hopefully_ be explanatory enough, but I'm not confident that this interface will help solve your concern

 A well-written comment on the implementation of `*runner` I feel would make navigating the code simpler, and allow us to put comments on the implementation functions instead of splitting the context between an interface.

Reply:
The specific problem I'm trying to solve is that I've got a type with some internal data that needs to be protected by a mutex/condition.  There were already some bugs in the current `provisionerd` relating to access of fields that were supposed to be protected by the mutex, where the code in question was not holding the lock.

Also, the runner has like a dozen methods for reasons of decomposing the work, most of which should only be called from the runners own methods.

Here I'm using the interface not to create a reusable abstraction, but to enlist the compiler/type checker in making sure that when the `Server` interacts with the running job, it doesn't break the invariants or call internal methods.  I agree that it's not the classic interface-as-abstraction use case that they teach you, but it gets the job done.  Like I said, if you're really opposed to an interface, I'll move this into a separate package and just expose the safe methods & data.

Reply:
I'm not convinced a future contributor would make the same assumptions about the purpose of this interface, I don't believe I intuitively would.

I think it'd be better to extract the runner into its own package. If you feel like that's too much for this PR, I'm fine with making an issue to do that in a future PR.